extensions/hid/led.m

```diff
@@ -0,0 +1,128 @@
+    deviceCFSetRef = IOHIDManagerCopyDevices(mgr);
+    if (!deviceCFSetRef)
+        goto Oops;
+    // how many devices in the set?
+    CFIndex deviceIndex, deviceCount = CFSetGetCount(deviceCFSetRef);
+    // allocate a block of memory to extact the device refs from the set into
+    refs = malloc(sizeof(IOHIDDeviceRef) * deviceCount);
+    if (!refs)
+        goto Oops;
+    // now extract the device refs from the set
+    CFSetGetValues(deviceCFSetRef, (const void **)(void *)refs);
+    // before we get into the device loop set up element matching dictionary
+    dic = _CreateMatchingDict(false, kHIDPage_LEDs, 0);
+    if (!dic)
+        goto Oops;
+    
+    for (deviceIndex = 0; deviceIndex < deviceCount; deviceIndex++)
+    {
+        // if this isn't a keyboard device...
+        if (!IOHIDDeviceConformsTo((IOHIDDeviceRef)refs[deviceIndex], kHIDPage_GenericDesktop,
+                                   kHIDUsage_GD_Keyboard))
+        {
+            continue;  // ...skip it
+        }
+        // copy all the elements
+        CFArrayRef elements = IOHIDDeviceCopyMatchingElements(refs[deviceIndex],
+                                                              (__bridge CFDictionaryRef)dic,
+                                                              kIOHIDOptionsTypeNone);
+        if(!elements)
+            continue;
+        // iterate over all the elements
+        CFIndex i, n = CFArrayGetCount(elements);
+        for (i = 0; i < n; i++)
+        {
+            IOHIDElementRef element = (IOHIDElementRef)CFArrayGetValueAtIndex(elements, i);
+            if(!element)
+                continue;
+            uint32_t usagePage = IOHIDElementGetUsagePage(element);
+            // if this isn't an LED element, skip it
+            if (kHIDPage_LEDs != usagePage) continue;
+            uint32_t elusage = IOHIDElementGetUsage(element);
+            if (elusage == usage)
+            {
+                // set led
+                //err = IOHIDDeviceOpen(refs[deviceIndex], 0);
+                err = 0;
+                if (err == 0) {
+                    // create the IO HID Value to be sent to this LED element
+                    uint64_t timestamp = 0;
+                    IOHIDValueRef val = IOHIDValueCreateWithIntegerValue(kCFAllocatorDefault,
+                                                                         element, timestamp,
+                                                                         target_value);
+                    if (val)
+                    {
+                        // now set it on the device
+                        IOHIDDeviceSetValue(refs[deviceIndex], element, val);
+                        CFRelease(val);
+                        success = true;
+                    }
+                    //IOHIDDeviceClose(refs[deviceIndex], 0);
+                }
+                break;
+            }
+        }
+        CFRelease(elements);
+    }
+    CFRelease(mgr);
+    
+Oops:
```

Comment:
`Oops` also needs to check if `mgr` is non-NULL and `CFRelease()` it, since there are a couple of places that could jump to `Oops` with `mgr` set.

Reply:
Thanks for the review! New commit pushed.